experimental/examples/token/Token.fst

```diff
@@ -0,0 +1,276 @@
+  transfer_maintains_supply stor.accounts add_from add_to num;
+  { total_supply = stor.total_supply;
+    accounts     = new_acc
+  }
+
+
+(*******************************************************************)
+(* End Type Definitions *)
+(**** Begin Validations On Tokens *)
+(*******************************************************************)
+
+
+val valid_transfer : token -> tx -> bool
+let valid_transfer token tx =
+  match tx.tx_data with
+  | Transfer {from_account; transfer_amount} ->
+    has_n token.storage.accounts from_account transfer_amount
+    && tx.tx_authroized_account = from_account
+  | Mint _ | Burn _ ->
+    false
+
+val valid_mint : token -> tx -> bool
+let valid_mint token tx =
+  match tx.tx_data with
+  | Mint mint           -> token.owner = tx.tx_authroized_account
+  | Transfer _ | Burn _ -> false
+
+val valid_burn : token -> tx -> bool
+let valid_burn token tx =
+  match tx.tx_data with
+  | Burn {burn_from_account; burn_amount} ->
+    has_n token.storage.accounts burn_from_account burn_amount
+    && tx.tx_authroized_account = burn_from_account
+  | Transfer _ | Mint _ ->
+    false
+
+
+(*******************************************************************)
+(* End validations on tokens *)
+(**** Begin Functions On Tokens *)
+(*******************************************************************)
+
+val token_transaction : (token -> tx -> Prims.Tot bool) -> Prims.Tot Type0
+let token_transaction f =
+  tok : token -> tx : tx { f tok tx } -> token
+
+val transfer : token_transaction valid_transfer
+let transfer token transaction =
+  match transaction.tx_data with
+  | Transfer {from_account; to_account; transfer_amount} ->
+    { token
+      with storage = transfer_stor token.storage
+                                   from_account
+                                   to_account
+                                   transfer_amount
+    }
+
+val mint : token_transaction valid_mint
+let mint token transaction =
+  match transaction.tx_data with
+  | Mint {mint_amount; mint_to_account} ->
+    transfer_add token.storage.accounts mint_to_account mint_amount;
+    { token
+      with storage = {
+        total_supply = token.storage.total_supply + mint_amount;
+        accounts     = account_add token.storage.accounts mint_to_account mint_amount
+      }}
+
+val burn : token_transaction valid_burn
+let burn token transaction =
+  match transaction.tx_data with
+  | Burn {burn_from_account; burn_amount} ->
+    transfer_sub token.storage.accounts burn_from_account burn_amount;
+    { token
+      with storage = {
+        total_supply = token.storage.total_supply - burn_amount;
+        accounts     = account_sub token.storage.accounts burn_from_account burn_amount
+      }}
+
+type transaction_error =
+  | Not_enough_funds
+  | Not_same_account
+  | Not_owner_token
+  | Not_enough_tokens
+
+val execute_transaction : token -> tx -> c_or transaction_error token
```

Comment:
We should prove properties of this top-level function

Reply:
That is done with the newest push